cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -127,27 +129,40 @@
    *
    * @param namespaceId  the namespace with schedules to be migrated
    * @param appMetaStore app metadata store with schedules to be migrated
+   * @param scheduler the old scheduler to get schedule status from
    * @return the lexicographically largest namespace id String with schedule migration completed
    */
-  public String migrateFromAppMetadataStore(NamespaceId namespaceId, Store appMetaStore) {
+  public String migrateFromAppMetadataStore(NamespaceId namespaceId, Store appMetaStore, Scheduler scheduler) {
+    LOG.info("Starting schedule migration for namespace '{}'", namespaceId);
     String completedNamespace = getMigrationCompleteNamespace();
     if (completedNamespace != null && completedNamespace.compareTo(namespaceId.toString()) > 0) {
+      LOG.info("Schedule migration has already been completed for namespace '{}'", namespaceId);
       return completedNamespace;
     }
     for (ApplicationSpecification appSpec : appMetaStore.getAllApplications(namespaceId)) {
       ApplicationId appId = namespaceId.app(appSpec.getName(), appSpec.getAppVersion());
       for (ScheduleSpecification scheduleSpec : appSpec.getSchedules().values()) {
         ProgramSchedule schedule = Schedulers.toProgramSchedule(appId, scheduleSpec);
         try {
-          addSchedule(schedule);
+          ProgramId programId = schedule.getProgramId();
+          ProgramScheduleStatus status = scheduler.scheduleState(programId, programId.getType().getSchedulableType(),
+                                                                 schedule.getName());
+          addScheduleWithStatus(schedule, status, System.currentTimeMillis());
         } catch (AlreadyExistsException e) {
           // This should never happen since no schedule with the same schedule key should exist before migration
-          LOG.warn("Schedule {} already exists before schedule migration", schedule.getScheduleId(), e);
+          LOG.warn("Schedule '{}' already exists before schedule migration. Skipped.", schedule.getScheduleId(), e);
+        } catch (NotFoundException e) {
+          LOG.warn("Schedule status of '{}' is not found during schedule migration. Skipped.",
+                   schedule.getScheduleId(), e);
```

Comment:
I don't think so. This can only happen when race condition happens and the schedule is already removed in the old scheduler. Just adding the schedule to the `ProgramScheduleStoreDataset` will not be triggered by any notification